### DISCUSSION TITLE
Discussing alternative ways to detect cards in current hand

### DISCUSSION THREAD

**Discussion by becurrie** [AdriaGual/marvel-snap-bot#97] [Ideas]
I thought it might be useful to create a discussion here to talk about some of the cons of the current implementation, and different ways we could improve the detection of your current hand within the emulator.

**How do things currently work?**
- As of now, we are using template matching through the OpenCV library in Python. This works okay, but has a lot of pitfalls based on how Marvel Snap displays cards in your hand.

**What are those cons?**
- Depending on how many cards are in your hand, the sizes of each card differs slightly, this throws a wrench into the search algorithm (even with a precision/threshold value applied) the bot still struggles to find images in every possible "spot" in a hand and will commonly miss cards.
- If you have enough energy to actually play a card, it is "brighter" in your hand, this also poses some problems with the bot.

**Some examples of these issues**
- I took the liberty of writing some unit tests to try and properly test the "current cards" utility in the bot, the pitfalls with the search algorithm show up pretty quickly.

On turn 1:
![turn_1](https://user-images.githubusercontent.com/24272608/221042247-83cd2944-c94e-48ec-ae75-fe6ee95368d2.png)
On turn 2:
![turn_2](https://user-images.githubusercontent.com/24272608/221042243-6cf7900d-a54c-402f-bcae-5b5716e0c9b3.png)
On turn 3:
![turn_3](https://user-images.githubusercontent.com/24272608/221042245-8b57ee52-2f2d-4fa1-bfdf-3ac8b193e3a8.png)

So in the examples above, we can see that as soon as the "template" being matched is no longer matching up to our image being searched, we run into issues.

**Current proposed solutions/enhancements**
- The most fully fleshed out approach is to use a proper computer vision algorithm through machine learning, @AdriaGual mentioned this in another discussion, we could use TensorFlow to do image classification. This is probably the best route to go, but involves the most work potentially.
- Another option could be us taking the "template" images available in the data directory, and upscaling/descaling the templates based on the percentage change whenever their are X cards in your current hand, I'm not sure how we'd figure out what that percentage is, but it might be an easier stop-gap to achieve before doing any proper ML work to find cards.

**Comment by paperc07** [comment]
Ok so how do we get started using TensorFlow, cause I went and pulled all images thinking it was best to have the actual legit images, but problem is bot can't read the images or else we'd be in there. But what can we do to start. 

Also with TensorFlow it collects images of cards from different angles and lighting conditions to create a dataset. So how do we get started? How do we test it to make sure it's picking up the cards, and last how do we intergrate it with what we got now or do we need to make a whole new bot?

Brace Face Baby#4938 is my discord

**Reply by becurrie** [reply]
Maybe @AdriaGual can speak more to this slightly? Have you worked with AI/Tensor at all? I unfortunately have not lol, but I imagine it would take a lot to get this running and working locally on a machine running the bot?

**Reply by paperc07** [reply]
maybe but he's been real busy lately. I have not, I don't even know how to code. If it wasn't for chat gpt I wouldn't have even gotten as far as I did lol.

**Reply by becurrie** [reply]
I'll probably take a dive into TensorFlow today and see if I can come up with anything, feel free to add me on Discord if you want as well: votum#3950

**Comment by becurrie** [comment]
There's also probably some ways to tinker with and/or change the way our template matching algorithm works, I'm going to look into using a pixel tolerance based algorithm that may ease some of the issues we run into with the card positions/sizes slightly changing from the original size of the templates used.

**Reply by becurrie** [reply]
This didn't pan out as well as I wanted it to, pixel falloff seemed to work better in some spots, but presented some false positives in other locations.

**Reply by paperc07** [reply]
damn so what else could we try

**Reply by becurrie** [reply]
I'll update my post here or create a new discussion with some of my findings through the TensorFlow work I'm doing now :) Depending on where I get with it anyways..

**Comment by becurrie** [comment]
fwiw, I think it's worth outlining some of the things we'll actually need to get TensorFlow working and integrated (also nice to keep things written down somewhere as I test some things).

**A dataset**
- I think we can get away with using the images that can be grabbed from the [Kaggle](https://www.kaggle.com/datasets/jeanmidev/marvel-snap-decks-and-cards) dataset, We likely would want some sort script or utility that can access this and find any cards we're missing and grab the art for them to be used as a reference image that will then be fed into a TensorFlow model.
- Once we have the images, we want to make sure they're formatted correctly to be fed in to TensorFlow as a training/validation dataset, I think the format here can look similar to what we do in the bot currently:

```
cards/
  cardname/
     cardname.png
```

**Augment the dataset**
- Descaling the reference images as well, this improves the TensorFlow performance, but also works out for the images we'll be searching for within the emulator.
- Since we only have one reference image per card in this case, we need to augment the images by applying different randomized transformations against each image, TensorFlow provides many useful tools to do this for us when training the dataset.

```python
data_augmentation = keras.Sequential(
    [
        layers.RandomFlip("horizontal", input_shape=(img_height, img_width, 3)),
        layers.RandomRotation(0.1),
        layers.RandomZoom(0.2),
        layers.RandomBrightness(0.5),
        layers.RandomContrast(0.5),
    ],
)
```

- This can be fed into our model directly so that as we train/fit the dataset, we use hundreds of versions of a reference image instead of the single reference image (this prevents "overfitting" from taking place within our model).

**Simple image classification**
- Once the model is trained, and the accuracy works for what we want to do, we'll need to convert the model to a TensorFlow Lite model and then that will be used in the actual bot here to perform predictions on images passed from the emulator.
- Our model file would be what we actually keep in the bot repository, then we can access and use it like so:

```python
interpreter = tf.lite.Interpreter(model_path=TF_MODEL_FILE_PATH)
classify_lite = interpreter.get_signature_runner('serving_default')

predictions_lite = classify_lite(sequential_1_input=img_array)['outputs']
score_lite = tf.nn.softmax(predictions_lite)
```

**Reply by paperc07** [reply]
So what sizes do the images need to be? So we can format them correctly, I hate how we can't just take the images from marvel snap zone lol.

**Reply by paperc07** [reply]
also I knew I saw that kaggle dataset somewhere before! I downloaded it a month ago but wasn't sure how we could use it.

**Reply by becurrie** [reply]
> So what sizes do the images need to be? So we can format them correctly, I hate how we can't just take the images from marvel snap zone lol.

I believe we can use the marvelsnapzone images, that is essentially what the "art" attribute within the Kaggle dataset links to, so we would be able to get the 1,024x1,024 cart art, then it's all about converting that information to a valid dataset and applying enough augmentations to have a decent sized set of training data.

**Comment by becurrie** [comment]
Some of the results I have so far look pretty promising... This is with a model trained with 300 augmented images from the deck I currently use, so there's 12 classes in total, and the training I did was pretty quick (10 epochs, 90 steps per epoch), on my machine it took about 5 minutes to run.

![accuracy_test_01](https://user-images.githubusercontent.com/24272608/221412460-0dde3ac5-679b-408a-a4ce-af1655f761aa.png)
![accuracy_test_02](https://user-images.githubusercontent.com/24272608/221412461-70170675-bcdc-400a-a894-69174cfee005.png)

One thing to note here is this is just a simple image classification model and requires the input images to be pretty apparent as to what they are, you can see from my input images, they're cropped images taken from my hand in game.

The bot currently crops a screenshot of the complete in game hand and matches templates against it.

Image detection with bounding boxes would be a better solution here I think, but just diving into ML/TensorFlow for the first time so I'm not sure what that will take.

**Possible solution to above problem**
- One thing I did think about was that in game, we currently have the ability to see how many cards are currently in the players hand by clicking on their player avatar on the top left, that will show how many cards are in our hand, we could then crop each known location and derive which cards are present that way.
- This would basically add two extra clicks during each turn performed by the bot, we would click once to open the profile, grab a screenshot and parse, then click again to close the prompt.

**Reply by paperc07** [reply]
This seems awesome and I would love to help but I don't know what I would need to be doing so if you wanna teach me or hit me up on discord I'll work on everything just lmk

**Comment by becurrie** [comment]
@paperc07 @AdriaGual 

Bit of an update here, I currently have a somewhat working prototype that is quite accurate when finding cards in the hand, here's some of those results.

Note: This isn't object detection where I pass a single image of multiple cards, but rather, me determining how many cards are in the hand, and then cropping screenshots of each known location a card would be, then predicting on that image. I've just applied the found image as a bounding box to the location where the cropped image was taken.

Images below show the course of two games, with the hand changing sizes and the expected cards still being found.

Added benefit here of us also being able to now track multiple of the same card in a hand at the same time.

Game 1:
![game_001_in_game_turn_1_test_result](https://user-images.githubusercontent.com/24272608/221584171-8502b5e0-95d2-42fc-b9c7-fd4f7e64f5c0.png)
![game_001_in_game_turn_2_test_result](https://user-images.githubusercontent.com/24272608/221584174-b6ee4e40-9f3b-4503-81af-f9ce90592b91.png)
![game_001_in_game_turn_3_test_result](https://user-images.githubusercontent.com/24272608/221584177-52ceb03e-6dde-49d7-abc4-d3f96dd8c4df.png)
![game_001_in_game_turn_4_test_result](https://user-images.githubusercontent.com/24272608/221584179-9a36b255-b2bd-46ff-8dfa-db99319fa38e.png)
![game_001_in_game_turn_5_test_result](https://user-images.githubusercontent.com/24272608/221584181-b89d9a01-270d-4830-a7ee-3bbe5dbe9661.png)
![game_001_in_game_turn_final_test_result](https://user-images.githubusercontent.com/24272608/221584187-b579bc16-c13f-4b29-8814-e237dcec156c.png)

Game 2:
![game_002_in_game_turn_1_test_result](https://user-images.githubusercontent.com/24272608/221584190-e79c0587-0958-4dec-a9c5-d2be2827f3aa.png)
![game_002_in_game_turn_2_test_result](https://user-images.githubusercontent.com/24272608/221584193-07205b6d-b271-4967-8745-43f14f354e5f.png)
![game_002_in_game_turn_3_test_result](https://user-images.githubusercontent.com/24272608/221584157-9d93f573-9682-493a-a079-88c8002ba257.png)
![game_002_in_game_turn_4_test_result](https://user-images.githubusercontent.com/24272608/221584161-08f10bcb-738d-4858-b0f6-7fe053e56dfd.png)
![game_002_in_game_turn_5_test_result](https://user-images.githubusercontent.com/24272608/221584164-05726fad-4319-48f9-b578-1c22759884ee.png)
![game_002_in_game_turn_final_test_result](https://user-images.githubusercontent.com/24272608/221584168-7b8e6c6c-a810-44a0-8789-27922e1f8100.png)

**Reply by paperc07** [reply]
Ok can you tell me how I could help or add me on discord?

**Reply by becurrie** [reply]
@paperc07 I've added you, my name is votum on Discord!

**Reply by paperc07** [reply]
@becurrie I don't see you what is your discord and ill add you, cause I just checked and I definitely don't have you

**Comment by AdriaGual** [comment]
@becurrie This looks amazing! would love to see the training part of your model and how to do that, not only for this project, but would be really helpful for the future ones, great work!

**Reply by becurrie** [reply]
votum#3950, add me on discord and we chat about some of the enhancements.

**Comment by rafaelchampion** [comment]
Hi everyone. I'm sorry, I'm by no means an expert in image recognition and classification, but thinking about how to detect cards, shouldn't it be more advisable to train the model based solely on the card logo? I reckon they are distinct enough to be trained. This would improve it even further since card variations use the same logo as the base models.

EDIT: [This ](https://www.reddit.com/r/MarvelSnap/comments/v576hc/here_ive_extracted_all_260_marvel_snap_card_logos/) Reddit thread has a large amount of the card logos with a transparent background (I'm not sure there's all of them)

**Reply by becurrie** [reply]
@rafaelchampion Great point, I'll try the link you provided and see what training results come up with :)
